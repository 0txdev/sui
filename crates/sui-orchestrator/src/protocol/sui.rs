// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{
    fs,
    net::{IpAddr, Ipv4Addr},
    path::PathBuf,
};

use move_core_types::account_address::AccountAddress;
use rand::{rngs::StdRng, SeedableRng};
use sui_config::{
    genesis::GenesisCeremonyParameters,
    genesis_config::{
        AccountConfig, GenesisConfig, ObjectConfigRange, ValidatorConfigInfo, ValidatorGenesisInfo,
    },
};
use sui_keys::keystore::{AccountKeystore, FileBasedKeystore};
use sui_types::{
    base_types::{ObjectID, SuiAddress},
    crypto::{AuthorityKeyPair, KeypairTraits, NetworkKeyPair, SuiKeyPair},
    multiaddr::Multiaddr,
};

use crate::{benchmark::BenchmarkParameters, client::Instance, settings::Settings};

use super::{ProtocolCommands, ProtocolMetrics};

/// All configurations information to run a sui client or validator.
pub struct SuiProtocol {
    /// The working directory on the remote hosts (containing the databases and configuration files).
    working_dir: PathBuf,
}

impl ProtocolCommands for SuiProtocol {
    const CLIENT_METRICS_PORT: u16 = 8081;

    fn protocol_dependencies() -> Vec<&'static str> {
        vec![
            // Install typical sui dependencies.
            "sudo apt-get -y install curl git-all clang cmake gcc libssl-dev pkg-config libclang-dev",
            // This dependency is missing from the Sui docs.
            "sudo apt-get -y install libpq-dev",
        ]
    }

    fn db_directories(&self) -> Vec<PathBuf> {
        let authorities_db = [&self.working_dir, &Self::AUTHORITIES_DB.into()]
            .iter()
            .collect();
        let consensus_db = [&self.working_dir, &Self::CONSENSUS_DB.into()]
            .iter()
            .collect();
        vec![authorities_db, consensus_db]
    }

    fn genesis_command<'a, I>(&self, instances: I) -> String
    where
        I: Iterator<Item = &'a Instance>,
    {
        let working_dir = self.working_dir.display();
        let ips = instances
            .map(|x| x.main_ip.to_string())
            .collect::<Vec<_>>()
            .join(" ");
        let genesis = [
            "cargo run --release --bin sui --",
            "genesis",
            &format!("-f --working-dir {working_dir} --benchmark-ips {ips}"),
        ]
        .join(" ");

        [
            &format!("mkdir -p {working_dir}"),
            "source $HOME/.cargo/env",
            &genesis,
        ]
        .join(" && ")
    }

    fn node_command<'a, I>(&self, instances: I) -> Box<dyn Fn(usize) -> String>
    where
        I: Iterator<Item = &'a Instance>,
    {
        let instances: Vec<_> = instances.cloned().collect();
        let listen_addresses = Self::make_listen_addresses(&instances);

        let working_dir = self.working_dir.clone();
        Box::new(move |i| {
            let validator_config = SuiProtocol::validator_configs(i);
            let config_path: PathBuf = [&working_dir, &validator_config.into()].iter().collect();
            let path = config_path.display();
            let address = listen_addresses[i].clone();

            let run = [
                "cargo run --release --bin sui-node --",
                &format!("--config-path {path} --listen-address {address}"),
            ]
            .join(" ");
            ["source $HOME/.cargo/env", &run].join(" && ")
        })
    }

    fn client_command<'a, I>(
        &self,
        _instances: I,
        parameters: &BenchmarkParameters,
    ) -> Box<dyn Fn(usize) -> String>
    where
        I: Iterator<Item = &'a Instance>,
    {
        let genesis_path: PathBuf = [&self.working_dir, &Self::GENESIS_BLOB.into()]
            .iter()
            .collect();
        let keystore_path: PathBuf = [&self.working_dir, &Self::GAS_KEYSTORE_FILE.into()]
            .iter()
            .collect();
        let committee_size = parameters.nodes;
        let load_share = parameters.load / committee_size;
        let shared_counter = parameters.shared_objects_ratio;
        let transfer_objects = 100 - shared_counter;
        let metrics_port = Self::CLIENT_METRICS_PORT;

        Box::new(move |i| {
            let genesis = genesis_path.display();
            let keystore = keystore_path.display();
            // let gas_id = SuiProtocol::gas_object_id_offsets(committee_size)[i].clone();
            let gas_id = GenesisConfig::benchmark_gas_object_id_offsets(committee_size)[i].clone();
            let run = [
                "cargo run --release --bin stress --",
                "--num-client-threads 24 --num-server-threads 1",
                "--local false --num-transfer-accounts 2",
                &format!("--genesis-blob-path {genesis} --keystore-path {keystore}",),
                &format!("--primary-gas-id {gas_id}"),
                "bench",
                &format!("--in-flight-ratio 30 --num-workers 24 --target-qps {load_share}"),
                &format!("--shared-counter {shared_counter} --transfer-object {transfer_objects}"),
                &format!("--client-metric-host 0.0.0.0 --client-metric-port {metrics_port}"),
            ]
            .join(" ");
            ["source $HOME/.cargo/env", &run].join(" && ")
        })
    }
}

impl ProtocolMetrics for SuiProtocol {
    const BENCHMARK_DURATION: &'static str = "benchmark_duration";
    const TOTAL_TRANSACTIONS: &'static str = "latency_s_count";
    const LATENCY_BUCKETS: &'static str = "latency_s";
    const LATENCY_SUM: &'static str = "latency_s_sum";
    const LATENCY_SQUARED_SUM: &'static str = "latency_squared_s";
}

impl SuiProtocol {
    /// Make a new instance of the Sui protocol commands generator.
    pub fn new(settings: &Settings) -> Self {
        Self {
            working_dir: [&settings.working_dir, &Self::CONFIG_DIR.into()]
                .iter()
                .collect(),
        }
    }

    /// Convert the ip of the validators' network addresses to 0.0.0.0.
    pub fn make_listen_addresses(instances: &[Instance]) -> Vec<Multiaddr> {
        let ips: Vec<_> = instances.iter().map(|x| x.main_ip.to_string()).collect();
        let genesis_config = GenesisConfig::new_for_benchmarks(&ips);
        let mut addresses = Vec::new();
        if let Some(validator_configs) = genesis_config.validator_config_info.as_ref() {
            for validator_info in validator_configs {
                let address = &validator_info.genesis_info.network_address;
                addresses.push(address.zero_ip_multi_address());
            }
        }
        addresses
    }
}

/// TODO: All these functions and variables are already defined in other parts of the codebase
/// or should not be needed after #9695 lands.
impl SuiProtocol {
    const AUTHORITIES_DB: &str = "authorities_db";
    const CONSENSUS_DB: &str = "consensus_db";
    const CONFIG_DIR: &str = "sui_config";
    const GENESIS_BLOB: &str = "genesis.blob";
    const GENESIS_CONFIG_FILE: &str = "benchmark-genesis.yml";
    pub const GAS_KEYSTORE_FILE: &str = "benchmark.keystore";

    pub fn validator_configs(i: usize) -> String {
        format!("validator-config-{i}.yaml")
    }

    fn gas_key() -> SuiKeyPair {
        let mut rng = StdRng::seed_from_u64(0);
        SuiKeyPair::Ed25519(NetworkKeyPair::generate(&mut rng))
    }

    pub fn gas_object_id_offsets(quantity: usize) -> Vec<String> {
        let mut rng = StdRng::seed_from_u64(0);
        (0..quantity)
            .map(|_| format!("{:#x}", ObjectID::random_from_rng(&mut rng)))
            .collect()
    }

    pub fn print_files(instances: &[Instance]) {
        let ips: Vec<_> = instances.iter().map(|x| x.main_ip.to_string()).collect();
        let genesis_config = GenesisConfig::new_for_benchmarks(&ips);
        // let genesis_config = Self::make_genesis_config(instances);
        let yaml = serde_yaml::to_string(&genesis_config).unwrap();
        let path = PathBuf::from(Self::GENESIS_CONFIG_FILE);
        fs::write(path, yaml).unwrap();

        let mut keystore = FileBasedKeystore::default();
        // let gas_key = Self::gas_key();
        let gas_key = GenesisConfig::benchmark_gas_key();
        keystore.add_key(gas_key).unwrap();
        keystore.set_path(&PathBuf::from(Self::GAS_KEYSTORE_FILE));
        keystore.save().unwrap();
    }

    pub fn configuration_files() -> Vec<PathBuf> {
        vec![
            Self::GENESIS_CONFIG_FILE.into(),
            // Self::GAS_KEYSTORE_FILE.into(),
        ]
    }

    pub fn genesis_config_command(&self, instances: &[Instance]) -> String {
        let ips = instances
            .iter()
            .map(|x| x.main_ip.to_string())
            .collect::<Vec<_>>()
            .join(" ");
        let genesis_file = format!("~/{}", Self::GENESIS_CONFIG_FILE);
        let genesis = [
            "cargo run --release --bin sui --",
            // &format!(
            //     "genesis -f --from-config {genesis_file} --working-dir {} --benchmark-ips {ips}",
            //     self.working_dir.display()
            // ),
            &format!(
                "genesis -f  --working-dir {} --benchmark-ips {ips}",
                self.working_dir.display()
            ),
        ]
        .join(" ");
        [format!("mkdir -p {}", self.working_dir.display()), genesis].join(" && ")
    }

    /// Generate a genesis configuration file suitable for benchmarks.
    pub fn make_genesis_config(instances: &[Instance]) -> GenesisConfig {
        let gas_key = Self::gas_key();
        let gas_address = SuiAddress::from(&gas_key.public());

        // Set the validator's configs.
        let mut rng = StdRng::seed_from_u64(0);
        let validator_config_info: Vec<_> = instances
            .iter()
            .map(|instance| {
                ValidatorConfigInfo {
                    consensus_address: "/ip4/127.0.0.1/tcp/8083/http".parse().unwrap(),
                    consensus_internal_worker_address: None,
                    genesis_info: ValidatorGenesisInfo::from_base_ip(
                        AuthorityKeyPair::generate(&mut rng), // key_pair
                        NetworkKeyPair::generate(&mut rng),   // worker_key_pair
                        SuiKeyPair::Ed25519(NetworkKeyPair::generate(&mut rng)), // account_key_pair
                        NetworkKeyPair::generate(&mut rng),   // network_key_pair
                        Some(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0))), // p2p_listen_address
                        instance.main_ip.to_string(),
                        500, // port_offset
                    ),
                }
            })
            .collect();

        // Generate the genesis gas objects.
        let genesis_gas_objects = Self::gas_object_id_offsets(instances.len())
            .iter()
            .map(|id| ObjectConfigRange {
                offset: AccountAddress::from_hex_literal(id).unwrap().into(),
                count: 5000,
                gas_value: 18446744073709551615,
            })
            .collect();

        // Set the initial gas objects.
        let account_config = AccountConfig {
            address: Some(gas_address),
            gas_objects: vec![],
            gas_object_ranges: Some(genesis_gas_objects),
        };

        // Make the genesis configuration file.
        GenesisConfig {
            validator_config_info: Some(validator_config_info),
            parameters: GenesisCeremonyParameters::new(),
            committee_size: instances.len(),
            grpc_load_shed: None,
            grpc_concurrency_limit: None,
            accounts: vec![account_config],
        }
    }
}
