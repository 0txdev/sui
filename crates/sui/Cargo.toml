[package]
name = "sui"
version.workspace = true
authors = ["Mysten Labs <build@mystenlabs.com>"]
license = "Apache-2.0"
publish = false
edition = "2021"

[dependencies]
json_to_table.workspace = true
tabled.workspace = true
anemo.workspace = true
anyhow.workspace = true
serde.workspace = true
serde_json.workspace = true
serde_yaml.workspace = true
signature.workspace = true
camino.workspace = true
tokio = { workspace = true, features = ["full"] }
async-trait.workspace = true
tracing.workspace = true
bcs.workspace = true
clap.workspace = true
bip32.workspace = true
rand.workspace = true
tap.workspace = true
inquire.workspace = true
rusoto_core.workspace = true
rusoto_kms.workspace = true
prometheus.workspace = true
git-version.workspace = true
const-str.workspace = true
num-bigint.workspace = true
regex.workspace = true
reqwest.workspace = true
im.workspace = true

sui-config.workspace = true
sui-execution = { path = "../../sui-execution" }
sui-swarm-config.workspace = true
sui-genesis-builder.workspace = true
sui-types.workspace = true
sui-json.workspace = true
sui-swarm.workspace = true
sui-json-rpc-types.workspace = true
sui-sdk.workspace = true
sui-keys.workspace = true
sui-source-validation.workspace = true
sui-move = { workspace = true, features = ["all"] }
sui-move-build.workspace = true
sui-protocol-config.workspace = true
shared-crypto.workspace = true
sui-replay.workspace = true

fastcrypto.workspace = true
fastcrypto-zkp.workspace = true

rustyline.workspace = true
rustyline-derive.workspace = true
colored.workspace = true
unescape.workspace = true
shell-words.workspace = true

tempfile.workspace = true
telemetry-subscribers.workspace = true

move-core-types.workspace = true
move-package.workspace = true
csv.workspace = true
workspace-hack.workspace = true

[target.'cfg(not(target_env = "msvc"))'.dependencies]
jemalloc-ctl.workspace = true

move-vm-test-utils = {workspace = true, optional = true}
move-vm-types = {workspace = true, optional = true}
move-vm-profiler = {workspace = true, optional = true}
move-cli = {workspace = true, optional = true}
move-unit-test = {workspace = true, optional = true}
move-vm-config = {workspace = true, optional = true}
sui-execution = {workspace = true, optional = true}
#move-vm-runtime = { path = "../../external-crates/move/move-execution/next-vm/crates/move-vm-runtime", option = true}

[dev-dependencies]
prometheus.workspace = true
fs_extra.workspace = true
expect-test.workspace = true
assert_cmd.workspace = true

test-cluster.workspace = true
sui-macros.workspace = true
sui-simulator.workspace = true
sui-test-transaction-builder.workspace = true

[package.metadata.cargo-udeps.ignore]
normal = ["jemalloc-ctl"]

[[example]]
name = "generate-genesis-checkpoint"
path = "src/generate_genesis_checkpoint.rs"
test = false

[[bin]]
name = "sui"
path = "src/main.rs"

[features]
default = ["gas-profiler"]
gas-profiler = [
    "sui-types/gas-profiler",
    "move-cli/gas-profiler",
    "move-unit-test/gas-profiler",
    "move-vm-config/gas-profiler",
    "move-vm-profiler/gas-profiler",
    "move-vm-types/gas-profiler",
    "move-vm-test-utils/gas-profiler",
    "sui-execution/gas-profiler",
    #"move-vm-runtime/gas-profiler",
]
