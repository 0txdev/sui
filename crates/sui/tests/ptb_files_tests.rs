// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeMap, path::Path};
use sui::client_ptb::{ptb::PTB, ptb_builder::errors::render_errors};
use sui_types::transaction::{CallArg, ObjectArg};
use test_cluster::TestClusterBuilder;

const TEST_DIR: &str = "tests";

#[cfg(not(msim))]
#[tokio::main]
async fn test_ptb_files(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let _ = miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .color(false)
                .width(80)
                .build(),
        )
    }));

    let fname = || path.file_name().unwrap().to_string_lossy().to_string();
    let file = path.to_str().unwrap().to_string();
    let args = vec!["--file".to_string(), file, "--preview".to_string()];
    let cwd = std::env::current_dir().unwrap();
    let mut commands = Vec::new();
    let ptb = PTB { args: args.clone() };
    ptb.parse_args(cwd, args, &mut BTreeMap::new(), &mut commands)?;

    // === PREVIEW ===
    let ptb_preview = ptb.preview(&commands.clone());
    let mut results = vec![];
    let preview_string = if let Some(ptb_preview) = ptb_preview {
        ptb_preview.to_string()
    } else {
        "".to_string()
    };
    results.push(" === PREVIEW === ".to_string());
    results.push(preview_string);

    // === PARSE COMMANDS ===
    let parsed_ptb_commands = ptb.parse_ptb_commands(commands.clone());
    match parsed_ptb_commands {
        Ok(ref commands) => commands,
        Err(e) => {
            // insta::assert_display_snapshot!(
            //     fname(),
            //     format!("=== FILE EXPANSION ERROR === \n {:?}", e)
            // );
            return Ok(());
        }
    };
    let parsed_ptb_commands = match parsed_ptb_commands {
        Ok(parsed) => parsed,
        Err(errors) => {
            let rendered = render_errors(&commands, errors);
            results.push(" === ERRORS AFTER PARSING INPUT COMMANDS === ".to_string());
            for e in rendered.iter() {
                results.push(format!("{:?}", e));
            }
            insta::assert_display_snapshot!(fname(), results.join("\n"));
            return Ok(());
        }
    };

    results.push(" === PARSED INPUT COMMANDS === ".to_string());

    for (_, c) in &parsed_ptb_commands {
        results.push(format!("cmd: {:?}", c));
    }

    // === BUILD PTB ===
    let test_cluster = TestClusterBuilder::new().build().await;
    let context = test_cluster.wallet;
    let client = context.get_client().await?;

    let built_ptb = ptb
        .parse_and_build_ptb(parsed_ptb_commands, &context, client)
        .await;

    if let Ok(ref ptb) = built_ptb {
        results.push(" === BUILT PTB === ".to_string());
        for (i, ca) in ptb.0.inputs.iter().enumerate() {
            results.push(format!("Input {}: {}", i, stable_call_arg_display(ca)));
        }
        for (i, c) in ptb.0.commands.iter().enumerate() {
            results.push(format!("Command {}: {}", i, c));
        }
    }

    // === BUILDING PTB ERRORS ===
    if let Err(e) = built_ptb {
        let rendered = render_errors(&commands, e);

        results.push(" === BUILDING PTB ERRORS === ".to_string());
        for e in rendered.iter() {
            results.push(format!("{:?}", e));
        }
    }

    // === FINALLY DO THE ASSERTION ===
    insta::assert_display_snapshot!(fname(), results.join("\n"));

    Ok(())
}

fn stable_call_arg_display(ca: &CallArg) -> String {
    match ca {
        CallArg::Pure(v) => format!("Pure({:?})", v),
        CallArg::Object(oa) => match oa {
            ObjectArg::ImmOrOwnedObject(_) => "ImmutableOrOwnedObject".to_string(),
            ObjectArg::SharedObject { mutable, .. } => {
                format!("SharedObject(mutable: {})", mutable)
            }
            ObjectArg::Receiving(_) => "Receiving".to_string(),
        },
    }
}

datatest_stable::harness!(test_ptb_files, TEST_DIR, r".*\.ptb$",);
