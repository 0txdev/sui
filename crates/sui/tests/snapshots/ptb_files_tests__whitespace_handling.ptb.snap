---
source: crates/sui/tests/ptb_files_tests.rs
expression: "results.join(\"\\n\")"
---
 === PREVIEW === 
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ PTB Preview                                                                                                                                                          │
├───────────────┬─────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ command       │ from                                            │ value(s)                                                                                           │
├───────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ assign        │ tests/ptb_files/parsing/whitespace_handling.ptb │ x vector[1,               2,       3,4                        ]                                    │
│ assign        │ tests/ptb_files/parsing/whitespace_handling.ptb │ x vector[1,               2,       3,4                        ]                                    │
│ assign        │ tests/ptb_files/parsing/whitespace_handling.ptb │ x vector['1',               '2,       3','','4                        ']                           │
│ assign        │ tests/ptb_files/parsing/whitespace_handling.ptb │ x vector['1',               '2,       3','','4                                    ']               │
│ make_move_vec │ tests/ptb_files/parsing/whitespace_handling.ptb │ <u16> [1,               2,       3,4                        ]                                      │
│ make_move_vec │ tests/ptb_files/parsing/whitespace_handling.ptb │ <sui::string::String> [1,               2,       3,4                        ]                      │
│ make_move_vec │ tests/ptb_files/parsing/whitespace_handling.ptb │ <sui::string::String> ['1',               '2,       3','','4                        ']             │
│ make_move_vec │ tests/ptb_files/parsing/whitespace_handling.ptb │ <sui::string::String> ['1',               '2,       3','','4                                    '] │
│ gas_budget    │ tests/ptb_files/parsing/whitespace_handling.ptb │ 100                                                                                                │
│ preview       │ console                                         │ true                                                                                               │
╰───────────────┴─────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────╯
 === PARSED INPUT COMMANDS === 
cmd: Assign(Spanned { span: Span { start: 0, end: 1, arg_idx: 0, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: "x" }, Some(Spanned { span: Span { start: 6, end: 61, arg_idx: 1, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 7, end: 8, arg_idx: 1, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(1) }, Spanned { span: Span { start: 24, end: 25, arg_idx: 1, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(2) }, Spanned { span: Span { start: 33, end: 34, arg_idx: 1, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(3) }, Spanned { span: Span { start: 35, end: 36, arg_idx: 1, file_scope: FileScope { file_command_index: 0, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(4) }]) }))
cmd: Assign(Spanned { span: Span { start: 0, end: 1, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: "x" }, Some(Spanned { span: Span { start: 6, end: 61, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 7, end: 8, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(1) }, Spanned { span: Span { start: 24, end: 25, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(2) }, Spanned { span: Span { start: 33, end: 34, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(3) }, Spanned { span: Span { start: 35, end: 36, arg_idx: 1, file_scope: FileScope { file_command_index: 1, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(4) }]) }))
cmd: Assign(Spanned { span: Span { start: 0, end: 1, arg_idx: 0, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: "x" }, Some(Spanned { span: Span { start: 6, end: 70, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 7, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'1'") }, Spanned { span: Span { start: 26, end: 38, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'2,       3'") }, Spanned { span: Span { start: 39, end: 41, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("''") }, Spanned { span: Span { start: 42, end: 69, arg_idx: 1, file_scope: FileScope { file_command_index: 2, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'4                        '") }]) }))
cmd: Assign(Spanned { span: Span { start: 0, end: 1, arg_idx: 0, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: "x" }, Some(Spanned { span: Span { start: 6, end: 82, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Vector([Spanned { span: Span { start: 7, end: 10, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'1'") }, Spanned { span: Span { start: 26, end: 38, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'2,       3'") }, Spanned { span: Span { start: 39, end: 41, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("''") }, Spanned { span: Span { start: 42, end: 81, arg_idx: 1, file_scope: FileScope { file_command_index: 3, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'4                                    '") }]) }))
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 5, arg_idx: 0, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U16 }, Spanned { span: Span { start: 0, end: 55, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 2, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(1) }, Spanned { span: Span { start: 18, end: 19, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(2) }, Spanned { span: Span { start: 27, end: 28, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(3) }, Spanned { span: Span { start: 29, end: 30, arg_idx: 1, file_scope: FileScope { file_command_index: 4, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(4) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 21, arg_idx: 0, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("sui"), name: "string" }, name: "String" }, type_args: [] }) }, Spanned { span: Span { start: 0, end: 55, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 2, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(1) }, Spanned { span: Span { start: 18, end: 19, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(2) }, Spanned { span: Span { start: 27, end: 28, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(3) }, Spanned { span: Span { start: 29, end: 30, arg_idx: 1, file_scope: FileScope { file_command_index: 5, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: U64(4) }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 21, arg_idx: 0, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("sui"), name: "string" }, name: "String" }, type_args: [] }) }, Spanned { span: Span { start: 0, end: 64, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'1'") }, Spanned { span: Span { start: 20, end: 32, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'2,       3'") }, Spanned { span: Span { start: 33, end: 35, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("''") }, Spanned { span: Span { start: 36, end: 63, arg_idx: 1, file_scope: FileScope { file_command_index: 6, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'4                        '") }] })
cmd: MakeMoveVec(Spanned { span: Span { start: 0, end: 21, arg_idx: 0, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: Struct(ParsedStructType { fq_name: ParsedFqName { module: ParsedModuleId { address: Named("sui"), name: "string" }, name: "String" }, type_args: [] }) }, Spanned { span: Span { start: 0, end: 76, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: [Spanned { span: Span { start: 1, end: 4, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'1'") }, Spanned { span: Span { start: 20, end: 32, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'2,       3'") }, Spanned { span: Span { start: 33, end: 35, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("''") }, Spanned { span: Span { start: 36, end: 75, arg_idx: 1, file_scope: FileScope { file_command_index: 7, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: String("'4                                    '") }] })
cmd: GasBudget(Spanned { span: Span { start: 0, end: 3, arg_idx: 0, file_scope: FileScope { file_command_index: 8, name: "tests/ptb_files/parsing/whitespace_handling.ptb", name_index: 0 } }, value: 100 })
cmd: Preview(Spanned { span: Span { start: 0, end: 4, arg_idx: 0, file_scope: FileScope { file_command_index: 1, name: "console", name_index: 0 } }, value: Bool(true) })
 === BUILT PTB === 
Input 0: Pure([1, 0, 0, 0, 0, 0, 0, 0])
Input 1: Pure([2, 0, 0, 0, 0, 0, 0, 0])
Input 2: Pure([3, 0, 0, 0, 0, 0, 0, 0])
Input 3: Pure([4, 0, 0, 0, 0, 0, 0, 0])
Input 4: Pure([3, 39, 49, 39])
Input 5: Pure([12, 39, 50, 44, 32, 32, 32, 32, 32, 32, 32, 51, 39])
Input 6: Pure([2, 39, 39])
Input 7: Pure([27, 39, 52, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 39])
Input 8: Pure([39, 39, 52, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 39])
Command 0: MakeMoveVec(Someu16,[Input(0),Input(1),Input(2),Input(3)])
Command 1: MakeMoveVec(Some0x2::string::String,[Input(0),Input(1),Input(2),Input(3)])
Command 2: MakeMoveVec(Some0x2::string::String,[Input(4),Input(5),Input(6),Input(7)])
Command 3: MakeMoveVec(Some0x2::string::String,[Input(4),Input(5),Input(6),Input(8)])
